client/relay_signer.go

```diff
@@ -0,0 +1,138 @@
+package client
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	"github.com/pokt-network/poktroll/pkg/crypto/rings"
+	"github.com/pokt-network/poktroll/x/application/types"
+	apptypes "github.com/pokt-network/poktroll/x/application/types"
+	servicetypes "github.com/pokt-network/poktroll/x/service/types"
+	"github.com/pokt-network/ring-go"
+)
+
+// publicKeyFetcher allows fetching the public key for a given address.
+// Satisfied by a full node, either with or without caching.
+type publicKeyFetcher interface {
+	getAccountPubKey(ctx context.Context, address string) (cryptotypes.PubKey, error)
+}
+
+// relaySigner holds the application or gateway's private key used to sign Relay Requests
+// and the public key fetcher to fetch the public key for the application.
+//
+// TODO_TECHDEBT(@commoddity): Investigate an alternative to requiring the public key fetcher.
+type relaySigner struct {
+	PrivateKeyHex    string
+	PublicKeyFetcher publicKeyFetcher
+}
+
+// SignRelayRequest signs the given relay request using the signer's private key and the application's ring.
+//
+// - Returns a pointer instead of directly setting the signature on the input relay request to avoid implicit output.
+// - Ideally, the function should accept a struct rather than a pointer, and also return an updated struct instead of a pointer.
+func (s *relaySigner) SignRelayRequest(
```

Comment:
This part looks incomplete: how is  a gateway (e.g. PATH) going to call this method?
SDK needs to provide this method, e.g. on the GatewayClient, for use by PATH.